src/backend/distributed/utils/node_metadata.c

```diff
@@ -0,0 +1,553 @@
+								   workerNodeLine),
+							errhint("Lines in the worker node file must contain a valid "
+									"node name and, optionally, a positive port number. "
+									"Comments begin with a '#' character and extend to "
+									"the end of their line.")));
+		}
+
+		/* allocate worker node structure and set fields */
+		workerNode = (WorkerNode *) palloc0(sizeof(WorkerNode));
+
+		strlcpy(workerNode->workerName, nodeName, WORKER_LENGTH);
+		workerNode->workerPort = nodePort;
+
+		workerNodeList = lappend(workerNodeList, workerNode);
+	}
+
+	FreeFile(workerFileStream);
+	free(workerFilePath);
+
+	/* rename the file, marking that it is not used anymore */
+	appendStringInfo(renamedWorkerFilePath, "%s", workerFilePath);
+	appendStringInfo(renamedWorkerFilePath, ".obsolete");
+	rename(workerFilePath, renamedWorkerFilePath->data);
+
+	return workerNodeList;
+}
+
+
+/*
+ * master_get_new_nodeid allocates and returns a unique nodeId for the node
+ * to be added. This allocation occurs both in shared memory and in write
+ * ahead logs; writing to logs avoids the risk of having nodeId collisions.
+ *
+ * Please note that the caller is still responsible for finalizing node data
+ * and the nodeId with the master node. Further note that this function relies
+ * on an internal sequence created in initdb to generate unique identifiers.
+ *
+ * NB: This can be called by any user; for now we have decided that that's
+ * ok. We might want to restrict this to users part of a specific role or such
+ * at some later point.
+ */
+Datum
+master_get_new_nodeid(PG_FUNCTION_ARGS)
+{
+	text *sequenceName = cstring_to_text(NODEID_SEQUENCE_NAME);
+	Oid sequenceId = ResolveRelationId(sequenceName);
+	Datum sequenceIdDatum = ObjectIdGetDatum(sequenceId);
+	Oid savedUserId = InvalidOid;
+	int savedSecurityContext = 0;
+	Datum shardIdDatum = 0;
+
+	GetUserIdAndSecContext(&savedUserId, &savedSecurityContext);
+	SetUserIdAndSecContext(CitusExtensionOwner(), SECURITY_LOCAL_USERID_CHANGE);
+
+	/* generate new and unique shardId from sequence */
+	shardIdDatum = DirectFunctionCall1(nextval_oid, sequenceIdDatum);
+
+	SetUserIdAndSecContext(savedUserId, savedSecurityContext);
+
+	PG_RETURN_DATUM(shardIdDatum);
+}
+
+
+/*
+ * master_get_next_groupid allocates and returns a unique groupId for the group
+ * to be created. This allocation occurs both in shared memory and in write
+ * ahead logs; writing to logs avoids the risk of having groupId collisions.
+ *
+ * Please note that the caller is still responsible for finalizing node data
+ * and the groupId with the master node. Further note that this function relies
+ * on an internal sequence created in initdb to generate unique identifiers.
+ *
+ * NB: This can be called by any user; for now we have decided that that's
+ * ok. We might want to restrict this to users part of a specific role or such
+ * at some later point.
+ */
+Datum
+master_get_next_groupid(PG_FUNCTION_ARGS)
```

Comment:
Minor: we have master_get_new_nodeid and master_get_next_groupid, should probably be both new or both next.